pandas/tests/reshape/merge/test_merge_anti.py

```diff
@@ -0,0 +1,502 @@
+                ),
+                "anti_full",
+                "B",
+                "B",
+            ),
+            (
+                DataFrame(
+                    {
+                        "A": [2, pd.NA],
+                        "B_x": [2, 3],
+                        "C": [np.nan, np.nan],
+                        "B_y": [np.nan, np.nan],
+                    }
+                ).astype({"B_x": object, "B_y": object}),
+                "anti_left",
+                "A",
+                "C",
+            ),
+            (
+                DataFrame(
+                    {
+                        "A": [np.nan, np.nan],
+                        "B_x": [np.nan, np.nan],
+                        "C": [1.0, 3],
+                        "B_y": [pd.NA, 2],
+                    }
+                ).astype(
+                    {
+                        "A": object,
+                        "B_x": object,
+                    }
+                ),
+                "anti_right",
+                "A",
+                "C",
+            ),
+            (
+                DataFrame(
+                    {
+                        "A": [2, pd.NA, np.nan, np.nan],
+                        "B_x": [2, 3, np.nan, np.nan],
+                        "C": [np.nan, np.nan, 1, 3],
+                        "B_y": [np.nan, np.nan, pd.NA, 2],
+                    }
+                ).astype({"B_x": object}),
+                "anti_full",
+                "A",
+                "C",
+            ),
+        ],
+    )
+    def test_anti_with_nan_and_NA(self, expected, how, left_on, right_on):
+        left = DataFrame({"A": [np.nan, 2, pd.NA], "B": ["a", 2, 3]})
+        right = DataFrame({"C": [1, 3, np.nan], "B": [pd.NA, 2, "c"]})
+        result = merge(left, right, how=how, left_on=left_on, right_on=right_on)
+        tm.assert_frame_equal(result, expected)
+
+    @pytest.mark.parametrize(
+        "how, expected",
+        [
+            (
+                "anti_left",
+                DataFrame(
+                    {"vals_x": [20, 17], "vals_y": [np.nan] * 2},
+                    index=pd.date_range("1/2/2010", periods=2, freq="2d"),
+                ),
+            ),
+            (
+                "anti_right",
+                DataFrame(
+                    {"vals_x": [np.nan] * 2, "vals_y": [17, 21]},
+                    index=pd.date_range("1/7/2010", periods=2, freq="2d"),
+                ),
+            ),
+            (
+                "anti_full",
+                DataFrame(
+                    {
+                        "vals_x": [20, 17, np.nan, np.nan],
+                        "vals_y": [np.nan, np.nan, 17, 21],
+                    },
+                    index=pd.date_range("1/7/2010", periods=2, freq="2d").union(
+                        pd.date_range("1/2/2010", periods=2, freq="2d")
+                    ),
+                ),
+            ),
+        ],
+    )
+    def test_anti_datetime(self, how, expected):
```

Comment:
same for datetime. It feels like you are testing everything everytime.
If you are confident from previous tests that the different "anti_xxx" work is there a sufficient reason to suggest that two might work and the other one not for a different dtype? 

If not then only one test is needed, not 3, and you might be able to combine with datetime_tz and make this much more readable.

Reply:
sure, I can test only one, and remove the other two.
I had separated the one with tz for readability (can combine into one if you think otherwise).

Reply:
my point being if you consider all the possible aspects to test:

dtypes: [float, int, bool, datetime, ea, categorical, etc.]
how: [anti_left, anti_right, anti_full]
kwargs: [on, left_on, right_on, left_index, right_index]
multiindex: [yes, no]
nan: [nan, pd.Na, None]

There is approximately over 500 different parametrised test combinations here. You seem to be doing around 50. I would try to get this down to about 25 or less, with structured logic.

For example, if you test that "how" works as expected for "anti_left", "anti_right" and "anti_full", and separately you show that "anti_full" works for different dtypes, you can make the logical assumption that for those additional dtypes "anti_left" and "anti_right" also work, so you dont have to test them.

I think it would be enough to show that pd.Na and nan and None are not treated as equal in a single test, and this would be enough to assume it works for all cases.

Reply:
I beg to differ here.
True that I have covered a small part of the complete set of combinations, but I don't agree that reducing that is a good idea. Different paths are taken at different places depending upon every param.
IOW just because something worked with single dtype wouldn't ensure it worked with mixed / EA

Reply:
OK, we probably share different philosophies. Mine is that pandas is a large package and if all possible developments were tested thoroughly with all params the tests might take a very long time. Ultimately the tests are there to prevent future regressions, they are not in place to test that everything works as developed: you can do that independently. Whilst possible, I would view it unlikely that someone introduces a regression which only breaks "anti_right" and only for one particular dtype. I think more likely it either breaks "anti_right" for all dtypes, or all the "anti_xx" combinations break for a single dtype. That is my logic, and why I would target those fewer tests.

But I acknowledge it is subjective and won't push it. :)